File: lib/python/src/test/opentoken/attributes/validation/birth_date_range_validator_test.py
"""
Tests for BirthDateRangeValidator.
"""

import pickle
from datetime import datetime, timedelta

import pytest

from opentoken.attributes.validation.birth_date_range_validator import BirthDateRangeValidator


class TestBirthDateRangeValidator:
    """Test cases for BirthDateRangeValidator."""

    def setup_method(self):
        """Set up test fixtures before each test method."""
        self.validator = BirthDateRangeValidator()

    def test_eval_valid_dates_within_range_should_return_true(self):
        """Test valid dates within the acceptable range."""
        # Test boundary dates that should be valid
        assert self.validator.eval("1910-01-01") is True  # Minimum valid date
        assert self.validator.eval("01/01/1910") is True  # Same date, different format

        # Test today's date (should be valid)
        today = datetime.now().date()
        assert self.validator.eval(today.strftime("%Y-%m-%d")) is True

        # Test dates within acceptable range
        assert self.validator.eval("1950-06-15") is True
        assert self.validator.eval("12/25/1990") is True
        assert self.validator.eval("29.02.2000") is True  # Leap year
        assert self.validator.eval("2000-12-31") is True
        assert self.validator.eval("01-01-2020") is True

    def test_eval_dates_before_minimum_should_return_false(self):
        """Test dates before the minimum valid date."""
        assert self.validator.eval("1909-12-31") is False
        assert self.validator.eval("12/31/1909") is False
        assert self.validator.eval("1900-01-01") is False
        assert self.validator.eval("01/01/1900") is False
        assert self.validator.eval("31.12.1909") is False
        assert self.validator.eval("1850-06-15") is False

    def test_eval_dates_after_today_should_return_false(self):
        """Test dates after today's date."""
        tomorrow = datetime.now().date() + timedelta(days=1)
        next_year = datetime.now().date() + timedelta(days=365)

        assert self.validator.eval(tomorrow.strftime("%Y-%m-%d")) is False
        assert self.validator.eval(next_year.strftime("%Y-%m-%d")) is False
        assert self.validator.eval("2030-01-01") is False
        assert self.validator.eval("01/01/2030") is False
        assert self.validator.eval("2050-12-25") is False

    def test_eval_invalid_date_formats_should_return_false(self):
        """Test invalid date formats."""
        assert self.validator.eval("20231026") is False  # No separators
        assert self.validator.eval("2023-13-01") is False  # Invalid month
        assert self.validator.eval("2023-02-30") is False  # Invalid day for February
        assert self.validator.eval("invalid-date") is False
        assert self.validator.eval("2023/15/45") is False  # Invalid month and day
        assert self.validator.eval("abc-def-ghi") is False

    def test_eval_null_and_empty_values_should_return_false(self):
        """Test null and empty values."""
        assert self.validator.eval(None) is False
        assert self.validator.eval("") is False
        assert self.validator.eval("   ") is False  # Whitespace only
        assert self.validator.eval("\t\n") is False

    def test_eval_various_date_formats_should_work_correctly(self):
        """Test all supported formats for the same date."""
        assert self.validator.eval("1995-07-15") is True  # yyyy-MM-dd
        assert self.validator.eval("1995/07/15") is True  # yyyy/MM/dd
        assert self.validator.eval("07/15/1995") is True  # MM/dd/yyyy
        assert self.validator.eval("07-15-1995") is True  # MM-dd-yyyy
        assert self.validator.eval("15.07.1995") is True  # dd.MM.yyyy

    def test_eval_leap_year_dates_should_work_correctly(self):
        """Test leap year dates."""
        # Valid leap year dates
        assert self.validator.eval("2000-02-29") is True  # Year 2000 is a leap year
        assert self.validator.eval("1996-02-29") is True  # 1996 is a leap year

        # Invalid leap year dates
        assert self.validator.eval("1900-02-29") is False  # 1900 is not a leap year
        assert self.validator.eval("2001-02-29") is False  # 2001 is not a leap year

    def test_serialization_should_preserve_validation_behavior(self):
        """Test serialization and deserialization of the validator."""
        # Serialize the validator
        serialized_data = pickle.dumps(self.validator)

        # Deserialize the validator
        deserialized_validator = pickle.loads(serialized_data)

        # Test that both validators behave identically
        test_values = [
            "1910-01-01",  # Valid boundary
            "1909-12-31",  # Invalid (too old)
            "2030-01-01",  # Invalid (future)
            "1990-06-15",  # Valid middle range
            "invalid-date",  # Invalid format
            None  # Null value
        ]

        for value in test_values:
            original_result = self.validator.eval(value)
            deserialized_result = deserialized_validator.eval(value)

            assert original_result == deserialized_result, (
                f"Validation results should match for value: {value} "
                f"(original: {original_result}, deserialized: {deserialized_result})"
            )

    def test_eval_edge_case_dates_should_work_correctly(self):
        """Test edge cases around boundaries."""
        today = datetime.now().date()
        yesterday = today - timedelta(days=1)
        min_date = datetime(1910, 1, 1).date()
        day_before_min_date = min_date - timedelta(days=1)

        # Test edge cases around boundaries
        assert self.validator.eval(today.strftime("%Y-%m-%d")) is True  # Today should be valid
        assert self.validator.eval(yesterday.strftime("%Y-%m-%d")) is True  # Yesterday should be valid
        assert self.validator.eval(min_date.strftime("%Y-%m-%d")) is True  # Minimum date should be valid
        assert self.validator.eval(day_before_min_date.strftime("%Y-%m-%d")) is False  # Day before minimum should be invalid

File: lib/java/src/test/java/com/truveta/opentoken/attributes/validation/BirthDateRangeValidatorTest.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes.validation;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.time.LocalDate;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class BirthDateRangeValidatorTest {

    private BirthDateRangeValidator validator;

    @BeforeEach
    void setUp() {
        validator = new BirthDateRangeValidator();
    }

    @Test
    void eval_ValidDatesWithinRange_ShouldReturnTrue() {
        // Test boundary dates that should be valid
        assertTrue(validator.eval("1910-01-01")); // Minimum valid date
        assertTrue(validator.eval("01/01/1910")); // Same date, different format

        // Test today's date (should be valid)
        LocalDate today = LocalDate.now();
        assertTrue(validator.eval(today.toString()));

        // Test dates within acceptable range
        assertTrue(validator.eval("1950-06-15"));
        assertTrue(validator.eval("12/25/1990"));
        assertTrue(validator.eval("29.02.2000")); // Leap year
        assertTrue(validator.eval("2000-12-31"));
        assertTrue(validator.eval("01-01-2020"));
    }

    @Test
    void eval_DatesBeforeMinimum_ShouldReturnFalse() {
        // Test dates before 1910-01-01
        assertFalse(validator.eval("1909-12-31"));
        assertFalse(validator.eval("12/31/1909"));
        assertFalse(validator.eval("1900-01-01"));
        assertFalse(validator.eval("01/01/1900"));
        assertFalse(validator.eval("31.12.1909"));
        assertFalse(validator.eval("1850-06-15"));
    }

    @Test
    void eval_DatesAfterToday_ShouldReturnFalse() {
        // Test dates after today
        LocalDate tomorrow = LocalDate.now().plusDays(1);
        LocalDate nextYear = LocalDate.now().plusYears(1);

        assertFalse(validator.eval(tomorrow.toString()));
        assertFalse(validator.eval(nextYear.toString()));
        assertFalse(validator.eval("2030-01-01"));
        assertFalse(validator.eval("01/01/2030"));
        assertFalse(validator.eval("2050-12-25"));
    }

    @Test
    void eval_InvalidDateFormats_ShouldReturnFalse() {
        // Test invalid date formats
        assertFalse(validator.eval("20231026")); // No separators
        assertFalse(validator.eval("2023-13-01")); // Invalid month
        assertFalse(validator.eval("2023-02-30")); // Invalid day for February
        assertFalse(validator.eval("invalid-date"));
        assertFalse(validator.eval("2023/15/45")); // Invalid month and day
        assertFalse(validator.eval("abc-def-ghi"));
    }

    @Test
    void eval_NullAndEmptyValues_ShouldReturnFalse() {
        assertFalse(validator.eval(null));
        assertFalse(validator.eval(""));
        assertFalse(validator.eval("   ")); // Whitespace only
        assertFalse(validator.eval("\t\n"));
    }

    @Test
    void eval_VariousDateFormats_ShouldWorkCorrectly() {
        // Test all supported formats for the same date
        assertTrue(validator.eval("1995-07-15")); // yyyy-MM-dd
        assertTrue(validator.eval("1995/07/15")); // yyyy/MM/dd
        assertTrue(validator.eval("07/15/1995")); // MM/dd/yyyy
        assertTrue(validator.eval("07-15-1995")); // MM-dd-yyyy
        assertTrue(validator.eval("15.07.1995")); // dd.MM.yyyy
    }

    @Test
    void eval_LeapYearDates_ShouldWorkCorrectly() {
        // Test valid leap year dates
        assertTrue(validator.eval("2000-02-29")); // Year 2000 is a leap year
        assertTrue(validator.eval("1996-02-29")); // 1996 is a leap year

        // Test invalid leap year dates
        assertFalse(validator.eval("1900-02-29")); // 1900 is not a leap year
        assertFalse(validator.eval("2001-02-29")); // 2001 is not a leap year
    }

    @Test
    void serialization_ShouldPreserveValidationBehavior() throws Exception {
        // Serialize the validator
        ByteArrayOutputStream byteOut = new ByteArrayOutputStream();
        ObjectOutputStream out = new ObjectOutputStream(byteOut);
        out.writeObject(validator);
        out.close();

        // Deserialize the validator
        ByteArrayInputStream byteIn = new ByteArrayInputStream(byteOut.toByteArray());
        ObjectInputStream in = new ObjectInputStream(byteIn);
        BirthDateRangeValidator deserializedValidator = (BirthDateRangeValidator) in.readObject();
        in.close();

        // Test that both validators behave identically
        String[] testValues = {
                "1910-01-01", // Valid boundary
                "1909-12-31", // Invalid (too old)
                "2030-01-01", // Invalid (future)
                "1990-06-15", // Valid middle range
                "invalid-date", // Invalid format
                null // Null value
        };

        for (String value : testValues) {
            boolean originalResult = validator.eval(value);
            boolean deserializedResult = deserializedValidator.eval(value);

            assertEquals(originalResult, deserializedResult,
                    String.format("Validation results should match for value: %s (original: %s, deserialized: %s)",
                            value, originalResult, deserializedResult));
        }
    }

    @Test
    void eval_EdgeCaseDates_ShouldWorkCorrectly() {
        LocalDate today = LocalDate.now();
        LocalDate yesterday = today.minusDays(1);
        LocalDate minDate = LocalDate.of(1910, 1, 1);
        LocalDate dayBeforeMinDate = minDate.minusDays(1);

        // Test edge cases around boundaries
        assertTrue(validator.eval(today.toString())); // Today should be valid
        assertTrue(validator.eval(yesterday.toString())); // Yesterday should be valid
        assertTrue(validator.eval(minDate.toString())); // Minimum date should be valid
        assertFalse(validator.eval(dayBeforeMinDate.toString())); // Day before minimum should be invalid
    }
}
